lib/topological_inventory/orchestrator/metric_scaler.rb

```diff
@@ -0,0 +1,186 @@
+        end
+
+        def start
+          @thread = Thread.new do
+            logger.info("Watcher thread for #{deployment_config_name} starting")
+            loop do
+              configure
+              break unless configured?
+
+              60.times do # Collect metrics for ~1 minute then check for config changes
+                moving_usage << percent_usage_from_metrics
+                sleep 1
+              end
+            end
+            logger.info("Watcher thread for #{deployment_config_name} stopping")
+          end
+        end
+
+        def stop
+          @thread.exit
+          @thread.join
+        end
+
+        def scale_to_desired_replicas
+          return unless configured?
+
+          desired_count = desired_replicas
+
+          return if desired_count == deployment_config.spec.replicas # already at max or minimum
+
+          logger.info("Scaling #{deployment_config_name} to #{desired_count} replicas")
+          object_manager.scale(deployment_config_name, desired_count)
+
+          # Wait for scaling to complete in Openshift
+          sleep(1) until pod_ips.length == desired_count
+          logger.info("Scaling #{deployment_config_name} complete")
+        end
+
+        private
+
+        def moving_usage
+          @moving_usage ||= FixedLengthArray.new(60)
+        end
+
+        def configure
+          logger.info("Fetching configuration for #{deployment_config_name}")
+          @deployment_config   = object_manager.get_deployment_config(deployment_config_name)
+          @current_metric_name = deployment_config.metadata.annotations["metric_scaler_current_metric_name"]       # i.e. "topological_inventory_api_puma_busy_threads"
+          @max_metric_name     = deployment_config.metadata.annotations["metric_scaler_max_metric_name"]           # i.e. "topological_inventory_api_puma_max_threads"
+          @max_replicas        = deployment_config.metadata.annotations["metric_scaler_max_replicas"]&.to_i        # i.e. "5"
+          @min_replicas        = deployment_config.metadata.annotations["metric_scaler_min_replicas"]&.to_i        # i.e. "1"
+          @target_usage_pct    = deployment_config.metadata.annotations["metric_scaler_target_usage_pct"]&.to_i    # i.e. "50"
+          @scale_threshold_pct = deployment_config.metadata.annotations["metric_scaler_scale_threshold_pct"]&.to_i # i.e. "20"
+        end
+
+        def desired_replicas
+          deviation = moving_usage.average.to_f - @target_usage_pct
+          count     = deployment_config.spec.replicas
+
+          return count if deviation.abs < @scale_threshold_pct # Within tolerance
+
+          deviation.positive? ? count += 1 : count -= 1
+          count.clamp(@min_replicas, @max_replicas)
+        end
+
+        def pod_ips
+          endpoint = object_manager.get_endpoint(deployment_config_name)
+          endpoint.subsets.flat_map { |s| s.addresses.collect { |a| a[:ip] } }
+        end
+
+        class FixedLengthArray
```

Comment:
So, normally I'm not particular pedantic about performance ... but this is going to be used constantly.

What you're looking for here is a [circular buffer](https://en.wikipedia.org/wiki/Circular_buffer). What you have implemented currently will create a new internal array object each time we insert a new value. It would be much better to use the same object internally.

[Here's](https://github.com/alindeman/circular_queue/blob/master/lib/circular_queue.rb) a gem that implements a queue with the concept, but the issue is that it assumes readers an writers (as is usually the case with a buffer queue). You would want to just drop the last item when enqueueing.

Reply:
I'm fine if you find something that accomplishes the task or implement your own, whatever seems best.